service/storage_proxy.cc

```diff
@@ -159,6 +165,66 @@ class storage_proxy::remote {
     netw::connection_drop_slot_t _connection_dropped;
     netw::connection_drop_registration_t _condrop_registration;
 
+    struct read_fence_params {
+        schema_ptr schema;
+        const dht::partition_range& range;
+    };
+
+    struct write_fence_params {
+        table_schema_version schema_version;
+        netw::msg_addr msg_from;
+        partition_key_view partition_key;
+    };
+
+    future<std::optional<replica::stale_topology_exception>> try_apply_fence(fencing_token fence,
+        lowres_clock::time_point timeout,
+        std::variant<write_fence_params, read_fence_params>&& params)
+    {
+        const auto current_version = _sp._shared_token_metadata.get()->get_topology_version();
+        if (fence.topology_version > current_version) {
+            throw std::runtime_error(format("fencing_token topology_version {} is greater "
+                                            "than the current topology version {}",
+                fence.topology_version, current_version));
+        }
+
+        if (fence.topology_version == current_version ||
+            (fence.allow_previous && (fence.topology_version + 1 == current_version)))
+        {
+            co_return std::nullopt;
+        }
+
+        schema_ptr schema;
+        dht::token token;
+        if (auto* wp = get_if<write_fence_params>(&params); wp) {
+            schema = co_await get_schema_for_write(wp->schema_version, wp->msg_from);
+            token = schema->get_partitioner().get_token(*schema, wp->partition_key);
+        } else {
+            auto&& rp = get<read_fence_params>(params);
+            schema = std::move(rp.schema);
+            token = rp.range.is_singular() ? rp.range.start()->value().token() : end_token(rp.range);
+        }
+        // it is possible that after the co_await get_schema token_metadata has
+        // already been rebuilt and got a larger topology_version than we observed
+        // before the outer if, but this does not affect the result of this function
+        const auto& ks_name = schema->ks_name();
+        const auto& ks = _sp._db.local().find_keyspace(ks_name);
+        auto erm = ks.get_effective_replication_map();
+        const auto& natural_endpoints = erm->get_natural_endpoints_without_node_being_replaced(token);
+        const auto& pending_endpoints = erm->get_token_metadata_ptr()->pending_endpoints_for(token, ks_name);
+
+        // Check if this replica is on the list of target replicas for mutation RPC.
+        // For read RPCs, if we are reading from the replica where the mutations go, it should be fine.
```

Comment:
Why do we need the check now? In new schema we wait for all previous requests to complete, so if a topology version check fails the request should be fenced.

Reply:
have no idea, entirely lost/demotivated in all this back-and-forth discussions

okay, will drop it

Reply:
dropped